shared/mathutil/math_helper.go

```diff
@@ -1,18 +1,32 @@
 package mathutil
 
-import "math"
+import (
+	"math"
+)
 
 // IntegerSquareRoot defines a function that returns the
-// largest possible integer root of a number.
+// largest possible integer root of a number using a divide and conquer
+// binary search approach:
+//
+// inspiration: https://www.geeksforgeeks.org/square-root-of-an-integer
 func IntegerSquareRoot(n uint64) uint64 {
-	x := n
-	y := (x + 1) / 2
-
-	for y < x {
-		x = y
-		y = (x + n/x) / 2
+	x := uint64(0)
+	y := uint64(1 << 32)
+	for {
+		if y <= 1+x {
+			return x
+		}
+		sqt := x + (y-x)/2
```

Comment:
You probably want to use `shr 1` instead of division by 2.

On a reasonable recent Haswell architecture i3/i5/i7-4XXX from 2013, the division instruction IDIV is 59x slower than a shift when argument is already in register (which is the case in the integer square root tight loop)

From https://www.agner.org/optimize/instruction_tables.pdf

![DeepinScreenshot_select-area_20191024151126](https://user-images.githubusercontent.com/22738317/67488888-dca3c080-f670-11e9-9fa9-198ddd87448c.png)

Reply:
Thanks for this! Excellent resource @mratsim will do